`Development`: Add application creation page

Douglas-Sarah opened the pull request:
<!-- Thanks for contributing to TumApply! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist

#### General

<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->

- [x] I tested **all** changes and their related features with **all** corresponding user types.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://confluence.aet.cit.tum.de/spaces/AP/pages/257786006/Language+Guidelines+Client).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311714/PR+Guidelines).

#### Server

- [x] **Important**: I implemented the changes with a [very good performance](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311718/Performance+Guidelines) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311877/Server+Guidelines).
- [x] I documented the Java code using JavaDoc style.

#### Client

- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311716/Client+Guidelines).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311925/Client+Test+Guidelines).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

This PR allows an Applicant to create an Application under the path `/application/create`.

### Description

<!-- Describe your changes in detail -->
- uses the Progressstepper to offer various input fields
- applicant can switch freely between the different panels and input their data
- applicant can save or sent the application at the end

### Steps for Testing

<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->

1. update database if necessary
2. build and run application
3. on homepage, click on Application Creation Button
4. test the application creation form
 
### Review Progress

<!-- Each PR should be reviewed by at least one other developers. The code, the functionality (= manual test). -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review

- [x] Code Review 1

#### Manual Tests

- [x] Test 1

### Screenshots

<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
**On the homepage, there is on the bottom a button that links to the application creation page**
![image](https://github.com/user-attachments/assets/b9e7a388-0c2b-40ae-be02-366bdec59db1)
Click on the button and you are navigated to the application creation page

**panel1**
![image](https://github.com/user-attachments/assets/01832e82-f6f9-497f-a0e4-f752ea1d93f9)

**panel2**
![image](https://github.com/user-attachments/assets/860c1bfa-7705-4a11-a03f-276ea83779fc)

**panel3**
![image](https://github.com/user-attachments/assets/3fe50173-14b4-443b-8d7d-7a98e62a809f)

MoSchmidt reviewed (approved):
I've tested the site locally, and it looks really good and everything works fine. I have left a few comments for the server code, but I assume this will be done later when we have authentication set up as well.
I also noticed that some options are in lower case “male” & “female”, while the countries are in upper case “Germany”. Is this intentional?

Except for that, the code looks really good👍

ani-ib commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.ts:
```diff
@@ -0,0 +1,257 @@
+    });
+  }
+
+  onPage1ValidityChanged(isValid: boolean): void {
+    this.page1Valid.set(isValid);
+  }
+
+  onPage2ValidityChanged(isValid: boolean): void {
+    this.page2Valid.set(isValid);
+  }
+
+  onPage3ValidityChanged(isValid: boolean): void {
+    this.page3Valid.set(isValid);
+  }
```
Thought for a future PR:  would it be possible to introduce a generic onValidityChanged method that takes the pageValid variable as a parameter? That way, instead of having three separate methods, we could reduce some of the boilerplate and make things a bit more streamlined

Douglas-Sarah replied:
That is a nice suggestion, I agree that it would make things more streamlined 👍

ani-ib commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+};
+
+export const dropdownGender: DropdownOption[] = [
+  { value: 'female', name: 'female' },
```
I would suggest making the values capitalised, to maintain consistency with the rest of the dropdown values.
```suggestion
  { value: 'Female', name: 'female' },
```

ani-ib commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+
+export const dropdownGender: DropdownOption[] = [
+  { value: 'female', name: 'female' },
+  { value: 'male', name: 'male' },
```
same here
```suggestion
  { value: 'Male', name: 'male' },
```

ani-ib commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+export const dropdownGender: DropdownOption[] = [
+  { value: 'female', name: 'female' },
+  { value: 'male', name: 'male' },
+  { value: 'other', name: 'other' },
```
same here
```suggestion
  { value: 'Other', name: 'other' },
```

ani-ib reviewed (changes requested):
Application Creation Page looks very nicely designed -
I had an issue with testing though - I cannot publish the Applications or save drafts. I get the following error in my server:
![image](https://github.com/user-attachments/assets/2e23a8ac-3d4c-4426-972a-29ea41e55394)
Could you look into this?
(Issue solved)

Douglas-Sarah commented on src/main/java/de/tum/cit/aet/application/domain/dto/CreateApplicationDTO.java:
```diff
@@ -1,18 +1,17 @@
-import java.util.Set;
 import java.util.UUID;
 
 public record CreateApplicationDTO(
```
Thanks for the review and the reminder- I added some validation 👍

Douglas-Sarah commented on src/main/java/de/tum/cit/aet/application/service/ApplicationService.java:
```diff
@@ -28,11 +35,38 @@ public ApplicationService(ApplicationRepository repository) {
+        applicant.setMasterGrade(createApplicationDTO.applicant().masterGrade());
+        applicant.setMasterUniversity(createApplicationDTO.applicant().masterUniversity());
+
+        Job job = jobRepository.getReferenceById(createApplicationDTO.jobId());
```
I added a check as well, that is a good point 👍

Douglas-Sarah commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+    city: application.applicant?.city ?? '',
+    country: application.applicant?.country ?? '',
+    postcode: application.applicant?.postalCode ?? '',
+    streetnumber: '', // TODO
```
We don't have the streetnumber in the database model yet, so I wanted to discuss if the streetnumber should be saved with the street and then separated, if it even is necessary to input them separately or if we need to adjust the schema.
For the review tomorrow I just build the model similar to the figma schema

Douglas-Sarah commented on src/main/webapp/app/pages/application/application-creation/application-creation-page2/application-creation-page2.component.scss:
```diff
@@ -0,0 +1,11 @@
+.form-row {
```
I agree in general. Maybe in this specific case we will even not need the scss in a global file when we have implemented a Form-Atomic-Component. But the concept is not thoroughly thought through (by myself at least)

Douglas-Sarah commented:
> Application Creation Page looks very nicely designed - I had an issue with testing though - I cannot publish the Applications or save drafts. I get the following error in my server: ![image](https://private-user-images.githubusercontent.com/88331280/444899754-2e23a8ac-3d4c-4426-972a-29ea41e55394.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc1ODQyNTYsIm5iZiI6MTc0NzU4Mzk1NiwicGF0aCI6Ii84ODMzMTI4MC80NDQ4OTk3NTQtMmUyM2E4YWMtM2Q0Yy00NDI2LTk3MmEtMjllYTQxZTU1Mzk0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTE4VDE1NTkxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI5ZmUyMjExZDZhMzUyMmY1ZGQ4ZGM0ZmE1ODRiNDIwNzM5ZTlkYTQxMDA4YTUwNGY5NTljODkyZWY2OTVlOTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.I48Cl5TEiBX-ZGv1kPLgxT6NdXQobofzj96FmVoxMTU) Could you look into this?

Thank you for your review and test! This is strange, I am looking into it 👍

Douglas-Sarah commented:
I have fixed the error : )

ani-ib reviewed (changes requested):
I've tested the code again and am now encountering another error in the server when attempting to save a draft or publish an application
<img width="1225" alt="image" src="https://github.com/user-attachments/assets/d9aec8d0-9211-44ae-a46d-9cd6a98e257f" />
(I've added the entire error messages below)

I also noticed that when you select a "Date of Birth" in the first page, then go to the next page and then come back to the first page, that the selected Date is not displayed anymore. This issue however does not happen to the "Desired Start Date" field. It would be nice if you could fix that 😊


Error messages:
2025-05-18T20:01:48.315+02:00  WARN 89037 --- [nio-8080-exec-7] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 1452, SQLState: 23000 
2025-05-18T20:01:48.315+02:00 ERROR 89037 --- [nio-8080-exec-7] o.h.engine.jdbc.spi.SqlExceptionHelper   : Cannot add or update a child row: a foreign key constraint fails (`tumapply`.`applications`, CONSTRAINT `fk_applications_job` FOREIGN KEY (`job_id`) REFERENCES `jobs` (`job_id`)) 
2025-05-18T20:01:48.322+02:00 ERROR 89037 --- [nio-8080-exec-7] d.t.c.a.c.e.h.GlobalExceptionHandler     : Handled exception: DataIntegrityViolationException - Path: /api/applications

Douglas-Sarah commented:
> I also noticed that when you select a "Date of Birth" in the first page, then go to the next page and then come back to the first page, that the selected Date is not displayed anymore. This issue however does not happen to the "Desired Start Date" field. It would be nice if you could fix that 😊

Thank you for noticing- I fixed the issue now : )

ani-ib commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+  postcode: string;
+};
+
+export const dropdownGender: DropdownOption[] = [
+  { value: 'Female', name: 'female' },
+  { value: 'Male', name: 'male' },
+  { value: 'Other', name: 'other' },
```
My previous code suggestion was mistaken, the name should be capitalized not the value
```suggestion
export const dropdownGender: DropdownOption[] = [
  { value: 'female', name: 'Female' },
  { value: 'male', name: 'Male' },
  { value: 'other', name: 'Other' },
```

ani-ib commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+  { value: 'Other', name: 'other' },
+];
+
+export const dropdownLanguage: DropdownOption[] = [
+  { value: 'de', name: 'german' },
+  { value: 'en', name: 'english' },
+];
```
I would capitalize the dropdown names here as well
```suggestion
export const dropdownLanguage: DropdownOption[] = [
  { value: 'de', name: 'German' },
  { value: 'en', name: 'English' },
];
```

ani-ib reviewed (changes requested):
Code looks good now, but my previous code suggestion had a mistake so I've added the correct code suggestion now

ani-ib reviewed (approved):
Code LGTM, thanks for implementing the changes 👍

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.ts:
```diff
@@ -0,0 +1,263 @@
+  private router = inject(Router);
+
+  constructor(private route: ActivatedRoute) {
+    this.route.url.subscribe(segments => {
```
please use async approaches over subscribe

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.ts:
```diff
@@ -0,0 +1,263 @@
+    });
+  }
+
+  ngOnInit(): void {
```
please do not use ngOnInit, rather use reactive approaches in modern angular. For example constructor with effects or computed

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.ts:
```diff
@@ -0,0 +1,263 @@
+          this.job = job;
+        });
+      } else {
+        this.router.navigate(['/404']);
```
please dont redirect to /404

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.ts:
```diff
@@ -0,0 +1,263 @@
+      const firstSegment = segments[1]?.path;
+      if (firstSegment === 'create') {
+        this.mode = 'create';
+        this.jobId = this.route.snapshot.paramMap.get('job_id')!;
```
do not use !

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.ts:
```diff
@@ -0,0 +1,263 @@
+    experiences: '',
+  };
+
+  @ViewChild('panel1', { static: true }) panel1!: TemplateRef<any>;
+  @ViewChild('panel2', { static: true }) panel2!: TemplateRef<any>;
+  @ViewChild('panel3', { static: true }) panel3!: TemplateRef<any>;
```
do not use !

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+  dropdownLanguageLocal = dropdownLanguage;
+  dropdownNationalityLocal = dropdownNationality;
+
+  constructor(private fb: FormBuilder) {}
```
please use inject instead of parameterized constructor injection

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+
+  constructor(private fb: FormBuilder) {}
+
+  ngOnInit(): void {
```
we should avoid ngOnInit here aswell if possible

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page2/application-creation-page2.component.html:
```diff
@@ -0,0 +1,78 @@
+<p-divider layout="horizontal">Bachelor</p-divider>
+<div class="form-row">
+  <div class="w-1/2 pr-4">
```
this styling can be generalized i would assume

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page2/application-creation-page2.component.ts:
```diff
@@ -0,0 +1,86 @@
+
+  form!: FormGroup;
+
+  constructor(private fb: FormBuilder) {}
```
same here use inject please instead of parameterized constructor injection

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page2/application-creation-page2.component.ts:
```diff
@@ -0,0 +1,86 @@
+
+  constructor(private fb: FormBuilder) {}
+
+  ngOnInit(): void {
```
same here avoid ngOninit

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page2/application-creation-page2.component.ts:
```diff
@@ -0,0 +1,86 @@
+  data = model.required<ApplicationCreationPage2Data>();
+  valid = output<boolean>();
+
+  form!: FormGroup;
```
dont use !

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page3/application-creation-page3.component.ts:
```diff
@@ -0,0 +1,57 @@
+
+  valid = output<boolean>();
+
+  page3Form!: FormGroup;
```
no ! please

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page3/application-creation-page3.component.ts:
```diff
@@ -0,0 +1,57 @@
+
+  page3Form!: FormGroup;
+
+  constructor(private fb: FormBuilder) {}
```
use inject() instead

Douglas-Sarah replied:
Changed it 👍

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page3/application-creation-page3.component.ts:
```diff
@@ -0,0 +1,57 @@
+
+  constructor(private fb: FormBuilder) {}
+
+  ngOnInit(): void {
```
avoid ngOnInit

Douglas-Sarah replied:
Changed it 👍

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page3/application-creation-page3.component.ts:
```diff
@@ -0,0 +1,57 @@
+      skills: [this.data().skills, Validators.required],
+    });
+
+    this.page3Form.valueChanges.subscribe(value => {
```
try the signal approach with effect and this part in the constructor

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+
+  valid = output<boolean>();
+
+  form!: FormGroup;
```
dont use !

Douglas-Sarah replied:
Changed it 👍

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page1/application-creation-page1.component.ts:
```diff
@@ -0,0 +1,156 @@
+      postcode: [this.data().postcode, Validators.required],
+    });
+
+    this.form.valueChanges.subscribe(value => {
+      Object.assign(this.data(), value);
+    });
+
+    this.form.statusChanges.subscribe(() => {
+      this.valid.emit(this.form.valid);
+    });
+
+    this.valid.emit(this.form.valid);
```
this can be done via signals in a modern way

Douglas-Sarah replied:
I am not sure how much this is possible with formGroup- this is a known issue.
See this issue: [Link to Angular Issue](https://github.com/angular/angular/issues/53485)

On the other hand, using FormGroups for validation used to be the best practice

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.ts:
```diff
@@ -0,0 +1,263 @@
+    experiences: '',
+  };
+
+  @ViewChild('panel1', { static: true }) panel1!: TemplateRef<any>;
```
angular has a signal version of this now viewChild() https://blog.angular-university.io/angular-17-2-release/

Douglas-Sarah replied:
changed it 👍

az108 reviewed (changes requested):
one comment i missed

Douglas-Sarah commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.html:
```diff
@@ -0,0 +1,23 @@
+<ng-template #panel1>
+  <div class="h-full flex p-1 m-4">
```
externalized into a scss class

Douglas-Sarah commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.ts:
```diff
@@ -0,0 +1,263 @@
+  stepData = signal<StepData[]>([]);
+  title?: string = 'Machine Learning for Climate Science';
+
+  jobId?: string;
```
I changed it for this PR, but for initiative applications it makes sense not to have a specific job.

Douglas-Sarah commented on src/main/webapp/app/pages/application/application-creation/application-creation-form/application-creation-form.component.ts:
```diff
@@ -0,0 +1,263 @@
+  jobId?: string;
+  job?: JobCardDTO;
+
+  mode?: ApplicationFormMode;
```
Changed it 👍

Douglas-Sarah commented:
I have not found a way to remove the subscribe- but the guy from artemis yesterday I talked to about this issue (I forgot his name) wants to look up if he found a way when he used form groups.
I will keep my eyes open.
I hope the rest is alright

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page2/application-creation-page2.component.ts:
```diff
@@ -0,0 +1,99 @@
+
+  fb = inject(FormBuilder);
+
+  constructor() {
```
i would prefer the approach above where we only use effect for the subscribe part and the subscribe part has an autoclean

az108 commented on src/main/webapp/app/pages/application/application-creation/application-creation-page3/application-creation-page3.component.ts:
```diff
@@ -0,0 +1,71 @@
+
+  fb = inject(FormBuilder);
+
+  constructor() {
```
same here

az108 commented on src/main/webapp/app/shared/components/atoms/string-input-temporary/string-input-temporary.component.ts:
```diff
@@ -0,0 +1,55 @@
+    },
+  ],
+})
+export class StringInputTemporaryComponent implements ControlValueAccessor {
```
we can remove this im merging the string input PR now

az108 reviewed (changes requested):
suggestion for the subscribe idea

Douglas-Sarah commented:
I implemented the requested changes and removed the string-input-temporary. Thanks for the suggestions, they were great.
In the future I might generalize the panels.

az108 reviewed (approved):
LGTM , thanks for implementing the changes suggested so quick 👍
